tools/infrasys/src/main.rs

```diff
@@ -0,0 +1,767 @@
+    let repos = infra_config
+        .repo
+        .as_mut()
+        .context(error::MissingConfig { missing: "repo" })?;
+
+    for (repo_name, repo_config) in repos.iter_mut() {
+        // Step 1: Create S3 Bucket
+        let bucket_name =
+            repo_config
+                .file_hosting_config_name
+                .as_ref()
+                .context(error::MissingConfig {
+                    missing: "file_hosting_config_name",
+                })?;
+        let mut s3_info = infra_config
+            .aws
+            .as_mut()
+            .context(error::MissingConfig { missing: "aws" })?
+            .s3
+            .get_mut(bucket_name)
+            .context(error::MissingConfig {
+                missing: format!("aws.s3 config with name {}", bucket_name),
+            })?;
+        let s3_region = s3_info.region.as_ref().context(error::MissingConfig {
+            missing: format!("region for '{}' s3 config", bucket_name),
+        })?;
+        let vpcid = s3_info
+            .vpc_endpoint_id
+            .as_ref()
+            .context(error::MissingConfig {
+                missing: format!("vpc_endpoint_id for '{}' s3 config", bucket_name),
+            })?;
+        // TODO: shouldn't be error, should have default
+        let prefix = s3_info.s3_prefix.as_ref().context(error::MissingConfig {
+            missing: format!("s3_prefix for '{}' s3 config", bucket_name),
+        })?;
+        let s3_stack_arn = create_s3_bucket(s3_region, bucket_name).await;
+        s3_info.stack_arn = Some(s3_stack_arn?);
+
+        // Step 2: Add Bucket Policy
+        // TODO: test adding bucket policy without the /*
+        add_bucket_policy(s3_region, bucket_name, vpcid, prefix);
+
+        // Step 3: Create root + publication keys
+        let signing_keys = repo_config
+            .signing_keys
+            .as_mut()
+            .context(error::MissingConfig {
+                missing: format!("signing_keys for '{}' repo config", repo_name),
+            })?;
+        let root_keys = repo_config
+            .root_keys
+            .as_mut()
+            .context(error::MissingConfig {
+                missing: format!("root_keys for '{}' repo config", repo_name),
+            })?;
+        create_keys(signing_keys).await; // TODO: add an ensure! ok check (make sure RESULT is used)
+        create_keys(root_keys).await;
+
+        // Step 4: Create and sign root.json
+        tuftool!(Region::default().name(), "root init '{}'", outdir.display());
+        tuftool!(
+            Region::default().name(),
+            // IN-FUTURE: expose expiration date as a configurable parameter
+            "root expire '{}' 'in 52 weeks'",
+            outdir.display()
+        );
+        // Root signing
+        sign_root(
+            get_available_keys(signing_keys)?, // QUESTION: What is the workflow of errors percolating up? Do I have to do any explicit error catching if the method()? resolves to error?
```

Comment:
If `get_available_keys` returns an error. The `?` here makes the current function return said error. So you're doing it correctly here. Unless you think it's ok for `get_available_keys` to error here, then you'd have to handle the error differently and not immediately return.